### ISSUE TITLE
Complex directive arguments

### ISSUE THREAD

**Issue by Hrle97** [graphql-dotnet/graphql-dotnet#1981]
## Summary

I was struggling to create a directive with complex arguments, so I made some workaround extension methods that make it possible to extract complex directive arguments the same way you would extract arguments for fields using the IResolveFieldContext interface. 
The downside is that you have to provide the schema to these methods and they have poor performance because they try to find the type information about the directive arguments straight from the schema. Also, they haven't been tested extensively, but I hope that they help someone struggling with the same issues.

## Relevant information

Here is the code:

    namespace GraphQL
    {
        public static class DirectiveExtensions
        {
            public static bool HasDirective(this IResolveFieldContext context, string name) =>
                context.GetDirective(name) != null;
                
            public static bool TryGetDirective(
                this IResolveFieldContext context,
                string                    name,
                out Directive?            directive)
            {
                switch (context.GetDirective(name))
                {
                    case null:
                        directive = null;
                        return false;
        
                    case { } result:
                        directive = result;
                        return true;
                }
            }
        
            public static Directive? GetDirective(this IResolveFieldContext context, string name) =>
                context.FieldAst
                       .Directives
                      ?.Find(name);
        
        
            public static bool HasArgument(
                this Directive directive,
                string         name) =>
                directive.GetArgument(name) != null;
        
            public static TArgument? GetRefArgument<TArgument>(
                this Directive directive,
                ISchema        schema,
                string         name,
                TArgument?     defaultValue = null)
                where TArgument : class =>
                directive.TryGetArgument(schema, typeof(TArgument), name, out var result) ?
                    result == null ?
                        defaultValue :
                        result as TArgument :
                    defaultValue;
        
            public static TArgument GetValueArgument<TArgument>(
                this Directive directive,
                ISchema        schema,
                string         name,
                TArgument      defaultValue = default)
                where TArgument : struct =>
                directive.TryGetArgument(schema, typeof(TArgument), name, out var result) ?
                    (TArgument?) result ?? defaultValue :
                    defaultValue;
        
            public static object? GetArgument(
                this Directive directive,
                ISchema        schema,
                Type           argumentType,
                string         name,
                object?        defaultValue = null) =>
                directive.TryGetArgument(schema, argumentType, name, out var result) ?
                    result == null && argumentType.IsValueType ?
                        defaultValue :
                        result :
                    defaultValue;
        
            public static bool TryGetArgument(
                this Directive directive,
                ISchema        schema,
                Type           argumentType,
                string         name,
                out object?    result)
            {
                switch (directive.GetArgument(name))
                {
                    case null:
                        result = null;
                        return false;
        
        
                    case IDictionary<string, object> inputObject when argumentType == typeof(object):
                        result = inputObject;
                        return true;
        
                    case IDictionary<string, object> _ when argumentType.IsPrimitive():
                        throw new InvalidOperationException(
                            $"Could not read primitive type '{argumentType.FullName}' " +
                            $"from complex argument '{name}'");
        
                    case IDictionary<string, object> inputObject:
                        result = inputObject.ToObject(
                            argumentType,
                            schema.FindArgumentType(directive.Name, name));
        
                        return true;
        
        
                    case { } arg:
                        result = arg.GetPropertyValue(
                            argumentType,
                            schema.FindArgumentType(directive.Name, name));
        
                        return true;
                }
            }
        
        
            private static object? GetArgument(this Directive directive, string name) =>
                directive
                   .Arguments
                   .ValueFor(name)
                  ?.Value;
        
        
            private static IGraphType FindArgumentType(this ISchema schema, string directiveName, string argumentName) =>
                schema
                   .Directives
                   .FirstOrDefault(
                        directive =>
                            directive.Name == directiveName)
                  ?.Arguments
                   .FirstOrDefault(
                        argument =>
                            argument.Name == argumentName)
                  ?.ResolvedType ??
                throw new InvalidOperationException(
                    $"{directiveName} argument {argumentName} not found on {schema}.");
        }
    }

**Comment by sungam3r** [comment]
Are you talking about directives supplied by client - http://spec.graphql.org/June2018/#ExecutableDirectiveLocation ?

**Comment by Hrle97** [comment]
Yes. Some queries of mine return a list of items and I needed to filter the lists based on values on some of the fields on the server-side of things (performance reasons) so I made a \@where query that takes in a list of fields and their values. The problem was extracting that list of fields and their values because I have been told on another issue that I should retrieve directive arguments from the FieldAST and as far as I understood the FieldAST only provides a list of directives which have a list of arguments which hold an object value, so I needed to coerce the type of the argument to a list of fields and their values, so I came up with this solution which was a bit of a copy and paste from the IResolveFieldContext.GetArgument method.

**Comment by sungam3r** [comment]
> I'm not sure what you mean by directives supplied by the client.

I mean directives in user provided query. @skip and @include are examples.

**Comment by Hrle97** [comment]
Ah right, that's it.

**Comment by Shane32** [comment]
@sungam3r This looks like something that could be added in 4.1 ?

**Comment by sungam3r** [comment]
Probably.

**Comment by sungam3r** [comment]
@Hrle97 Feature will be released in v5.
